client/asset/dcr/dcr.go

```diff
@@ -3078,6 +3091,296 @@ func (dcr *ExchangeWallet) EstimateRegistrationTxFee(feeRate uint64) uint64 {
+// message, sha256(account ID). This is the Bond.BondAcctSig field, which the
+// server should verify using the account ID and pubkey recovered from the
+// signature via ecdsa.RecoverCompact(sig, msg). Bond.RedeemTx is a backup
+// transaction that spends the bond output after lockTime passes, paying to an
+// address for the current underlying wallet; the bond private key (BondPrivKey)
+// should normally be used to author a new transaction paying to a new address
+// instead.
+//
+// NOTE: Alternatively, the caller could provide the pubKey, where the private
+// key is managed by the caller instead of the wallet. The caller would also
+// create the BondAcctSig using a signature algorithm and message (e.g. acct ID
+// hash) of their choice.
+func (dcr *ExchangeWallet) MakeBondTx(ver uint16, amt uint64, lockTime time.Time, acctID []byte) (*asset.Bond, error) {
+	if ver != 0 {
+		return nil, errors.New("only version 0 bonds supported")
+	}
+	if until := time.Until(lockTime); until >= 365*12*time.Hour /* ~6 months */ {
+		return nil, fmt.Errorf("that lock time is nuts: %v", lockTime)
+	} else if until < 0 {
+		return nil, fmt.Errorf("that lock time is already passed: %v", lockTime)
+	}
+
+	// Get an address to which the time-locked fidelity bond should pay. This
+	// address' keys are also used to sign a message (sha256 of account ID) for
+	// the server to verify the acct owns this bond. NOTE: Instead of a wallet
+	// key pair for the bond script, we could pay to the account pubkey, but
+	// that would be poor security.
+	addr, err := dcr.wallet.InternalAddress(dcr.ctx, dcr.config().primaryAcct)
+	if err != nil {
+		return nil, fmt.Errorf("error getting address for TL fidelity bond output: %w", translateRPCCancelErr(err))
+	}
+	if _, ok := addr.(*stdaddr.AddressPubKeyHashEcdsaSecp256k1V0); !ok {
+		return nil, fmt.Errorf("wallet did not return a v0 p2pkh, got a %T", addr)
+	}
+	priv, err := dcr.wallet.AddressPrivKey(dcr.ctx, addr)
+	if err != nil {
+		return nil, err
+	}
+	pk := priv.PubKey().SerializeCompressed()
+
+	baseTx := wire.NewMsgTx()
+	const scriptVersion = 0
+
+	// TL output.
+	bondScript, err := txscript.NewScriptBuilder().
+		AddInt64(lockTime.Unix()).
+		AddOp(txscript.OP_CHECKLOCKTIMEVERIFY).
+		AddOp(txscript.OP_DROP).
+		AddData(pk[:]).
+		AddOp(txscript.OP_CHECKSIG).
+		Script()
+	if err != nil {
+		return nil, fmt.Errorf("failed to build bond output redeem script: %w", err)
+	}
+	bondAddr, err := stdaddr.NewAddressScriptHash(scriptVersion, bondScript, dcr.chainParams)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build bond output payment script: %w", err)
+	}
+	bondPkScriptVer, bondPkScript := bondAddr.PaymentScript()
+	txOut := newTxOut(int64(amt), bondPkScriptVer, bondPkScript)
+	baseTx.AddTxOut(txOut)
+
+	// Acct ID commitment output
+	commitPkScript, err := txscript.NewScriptBuilder().
+		AddOp(txscript.OP_RETURN).
+		AddData(acctID).
```

Comment:
Bond version number somewhere here?

Reply:
https://github.com/decred/dcrdex/pull/1819#issuecomment-1238241718
Includes the whole shebang now.